### PR TITLE
Python: rewrite as a tokenizing parser (rebased version)

### PR DESCRIPTION
**Pull request by b4n**
This is the rebased (cleaned up history) version of #1003.

It has a very slight difference: the indentation of *Units/parser-python.r/bug1809024.py.d/input.py* is different, because I didn't re-change it and kept the version currently in master (in the other PR, I used the version I fixed myself).

The other difference is obviously the history.

I can leave it like that, or further squash it.  This state is as incremental as possible (well, easily at least) without introducing regressions in the middle of the way.

### CODE REVIEW

File: Units/parser-python.r/newlines-cr.d/README [hunk 1]
```diff
@@ -0,0 +1,2 @@
+Tags are right, but location are not recorded correctly.
+Likely due to a bug in main part line number recording with CRs alone.
```
Comment (masatake) [inline Units/parser-python.r/newlines-cr.d/README]: In tthat case, could you change the suffix of this directory to `.b` from `.d`?
So we can remember there is a bug.

**Comment by masatake** [comment]
I'm very sorry but could you remove `04dde62354b92496d0c45811ac06cbd45586939c`?

I asked add the empty decoration fields. However,  I feel they are noisy after reading real tags output which includes empty decoration: fields.

**Comment by masatake** [comment]
Other than two things I pointed out(about `.b` suffix and the empty decorator fields), Looks good.
Could you merge these?

File: Units/parser-python.r/newlines-cr.d/README (hunk 1 above)
Comment (b4n) [inline Units/parser-python.r/newlines-cr.d/README]: OK

**Comment by b4n** [comment]
> I asked add the empty decoratorfields. However, I feel they are noisy after reading real tags output which includes empty decorator: fields.

Should I then do the same for inheritance?

File: Units/parser-python.r/newlines-cr.d/README (hunk 1 above)
Comment (b4n) [inline Units/parser-python.r/newlines-cr.d/README]: done

**Comment by masatake** [comment]
Thank you for updating.

> Should I then do the same for inheritance?

Sorry, but I don't understand the "same" in this context.
`fieldSpec` of "inheritance" field has `isValueAvailable` method.
With this method the main part can know whether the value should be rendered or not.

If you are talking about this method, the answer is no because this method can be used only in common fields; decorator field is not common field.

**Comment by b4n** [comment]
> Sorry, but I don't understand the "same" in this context.

I mean, should I make the parser write the `inheritance` field only when it found something to put there, or should it leave it unset when there's nothing to put there.

But your answer suggests that there is a mechanism in core for hiding the field in some situations, so maybe it's not important.

**Comment by masatake** [comment]
> I mean, should I make the parser write the inheritance field only when it found something to put there, or should it leave it unset when there's nothing to put there.

Oh, I see.

O.k. I have no more comment. Could you merge this? Let's get feedbacks from real users.

**Comment by masatake** [comment]
Thank you.

**Comment by tek** [comment]
this seems to have removed the possiblity to omit imports as tags. With `--python-kinds=-Ix`, i get a tag for `Foo` at `from mod import Foo`, which was previously ignored with `--python-kinds=-i`.

**Comment by b4n** [comment]
@tek AFAIK this didn't change from the previous parser as it was in master: the `Foo` you see generated a tag of `unknown` kind, so you'd have to disable it with `--kinds-python=-x`.

If it's indeed a change from the *latest version* of the previous parser, please open a separate report.

**Comment by tek** [comment]
never mind – it seems I had updated ctags yesterday before this PR, updating again now fixed the issue: `--python-kinds=-x` works as expected.

Anyway, there is still a tag created for the module from which is being imported; while I have no quarrel with that, it seems to be disabled by `--python-kinds=-i`, which is not listed by `--list-kinds=python`. Thought that might be a mistake.

**Comment by b4n** [comment]
@tek Hum.  Indeed there is a difference between current master and previous pareser's results.  Apparently reference tags are always emitted, even without `--extra=+r` -- and even with `--extra=-r`.

I however am really not sure what's doing this… @masatake?

**Comment by b4n** [comment]
#1060

**Comment by b4n** [comment]
@tek behavior regarding reference tags should now be fixed.

**Comment by tek** [comment]
looks good!

**Pull request merged**